primitives/src/v0.rs

```diff
@@ -951,69 +950,172 @@ impl<Payload: EncodeAs<RealPayload>, RealPayload: Encode> Signed<Payload, RealPa
 		validator_index: ValidatorIndex,
 		key: &ValidatorId,
 	) -> Result<Option<Self>, KeystoreError> {
-		let data = Self::payload_data(&payload, context);
-		let signature = CryptoStore::sign_with(
-			&**keystore,
-			ValidatorId::ID,
-			&key.into(),
-			&data,
-		).await?;
+		let r = UncheckedSigned::sign(keystore, payload, context, validator_index, key).await?;
+		Ok(r.map(Self))
+	}
 
-		let signature = match signature {
-			Some(sig) => sig.try_into().map_err(|_| KeystoreError::KeyNotSupported(ValidatorId::ID))?,
-			None => return Ok(None),
-		};
+	/// Try to convert from `UncheckedSigned` by checking the signature.
+	pub fn from_unchecked<H: Encode>(
```

Comment:
Because it needs additional data (signing context and `ValidatorId`) to do the "conversion". But the `try_from_unchecked` is more consistent - will change that.